File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/telemetry/UnitPresentation.kt
package expo.modules.vescapecore.telemetry

/**
 * @parity /modules/vescape-core/ios/telemetry/UnitPresentation.swift
 * @parity /src/helpers/units.ts
 */
object UnitPresentation {
    const val METERS_PER_MILE = 1609.344
    fun speedFromKmh(kmh: Double, unitSystem: String): Double =
        if (unitSystem == "imperial") kmh * 1000.0 / METERS_PER_MILE else kmh
    fun speedUnit(unitSystem: String): String = if (unitSystem == "imperial") "mph" else "km/h"
    fun formatSpokenSpeed(kmh: Double, unitSystem: String): String =
        String.format(java.util.Locale.US, "%.0f", speedFromKmh(kmh, unitSystem))
    fun distance(meters: Double, unitSystem: String): String {
        if (!meters.isFinite()) return "—"
        val imperial = unitSystem == "imperial"
        val short = if (imperial) meters / METERS_PER_MILE < 0.1 else meters < 1000
        val value = if (imperial) meters / METERS_PER_MILE * (if (short) 5280 else 1) else meters / (if (short) 1 else 1000)
        val unit = if (imperial) { if (short) "ft" else "mi" } else { if (short) "m" else "km" }
        return String.format(java.util.Locale.US, if (short) "%.0f %s" else "%.1f %s", value, unit)
    }
}

File: modules/vescape-core/ios/telemetry/UnitPresentation.swift
import Foundation

/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/telemetry/UnitPresentation.kt
/// @parity /src/helpers/units.ts
internal enum UnitPresentation {
  static let metersPerMile = 1609.344
  static func speedFromKmh(_ kmh: Double, _ unitSystem: String) -> Double {
    unitSystem == "imperial" ? kmh * 1000 / metersPerMile : kmh
  }
  static func speedUnit(_ unitSystem: String) -> String { unitSystem == "imperial" ? "mph" : "km/h" }
  static func formatSpeed(_ kmh: Double, _ unitSystem: String) -> String {
    let value = String(format: "%.1f", locale: Locale(identifier: "en_US_POSIX"), speedFromKmh(kmh, unitSystem))
    return value.hasSuffix(".0") ? String(value.dropLast(2)) : value
  }
  static func distance(_ meters: Double, unitSystem: String) -> String {
    guard meters.isFinite else { return "—" }
    let imperial = unitSystem == "imperial"
    let short = imperial ? meters / metersPerMile < 0.1 : meters < 1000
    let value = imperial ? meters / metersPerMile * (short ? 5280 : 1) : meters / (short ? 1 : 1000)
    let unit = imperial ? (short ? "ft" : "mi") : (short ? "m" : "km")
    let scale = short ? 1.0 : 10.0
    return String(format: short ? "%.0f %@" : "%.1f %@", locale: Locale(identifier: "en_US_POSIX"), (value * scale).rounded(.toNearestOrAwayFromZero) / scale, unit)
  }
}
